VSR: Checkpoint after compact

sentientwaffle opened the pull request:
Checkpoint the `Forest` and then `SuperBlock` every `config.lsm_batch_multiple` ops.

Added `op_ceiling()` to hopefully avoid off-by-one errors when using `op_checkpoint` to compute the maximum op that can be received.
`op_checkpoint` starts at 0, then goes to 3, 7, 11, etc.

Also includes an assertion fix required for WAL wrapping.

Untested.

sentientwaffle committed:
VSR: Checkpoint after compact

Checkpoint the `Forest` and then `SuperBlock` every `config.lsm_batch_multiple` ops.

Added `op_ceiling()` to hopefully avoid off-by-one errors when using `op_checkpoint` to compute the maximum op that can be received.
`op_checkpoint` starts at 0, then goes to 3, 7, 11, etc.

Also includes a miscellaneous assertion fix required for WAL wrapping.

Untested.

sentientwaffle committed:
VSR: Checkpoint fixes; check invariants

- Initialize the `Replica`'s `view`, `view_normal, `commit_min`, `commit_max`, and `op_checkpoint` from the `SuperBlock`.
- Enforce and explain config restrictions around `lsm_batch_multiple`/`journal_slot_count`.

jorangreef commented on src/config.zig:
```diff
@@ -302,7 +302,24 @@ comptime {
+    //   A    B    C    D    E
+    //   |····|····|····|····|
+    //
+    // - ("|" delineates measures/batches; "·" is a prepare in the WAL.)
```
Nice docs!

sentientwaffle committed:
VSR/StateMachine: Hook up compact

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -2388,10 +2399,75 @@ pub fn Replica(
+            assert(self.committing);
+            assert(self.commit_callback != null);
+            assert(self.commit_prepare.?.header.op == self.commit_min);
+            assert((self.commit_prepare.?.header.op + 1) % config.journal_slot_count == 0);
```
I don't know if it's better but we could also get the slot for the op, to be able to compare it `journal_slot_count` in terms of the same unit?

jorangreef commented on src/config.zig:
```diff
@@ -302,7 +302,24 @@ comptime {
+    //   A    B    C    D    E
+    //   |····|····|····|····|
+    //
+    // - ("|" delineates measures/batches; "·" is a prepare in the WAL.)
```
```suggestion
    // - ("|" delineates measures, where a measure is a multiple of prepare batches; "·" is a prepare in the WAL.)
```

sentientwaffle replied:
Thanks! I was misunderstanding `lsm_batch_multiple` as "the cardinality of the batch" rather the "the number of batches in a measure".

jorangreef replied:
We should find a better name for it!

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -267,17 +268,15 @@ pub fn Replica(
             var clock = try Clock.init(
                 allocator,
                 replica_count,
-                replica,
+                replica_index,
```
Much better name, lines up too.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -2388,10 +2399,75 @@ pub fn Replica(
+            //   |01|23|45|67|
+            //
+            // The checkpoint is triggered at "E".
+            // The but only "A…D" is checkpointed, so the SuperBlock's `commit_min` is 7-2=5.
```
```suggestion
            // The checkpoint is triggered at "E".
            // At this point, ops 6 and 7 are in the in-memory immutable table.
            // They will only be compacted to disk in the next measure.
            // Therefore, only ops "A..D" are committed to disk.
            // Thus, the SuperBlock's `commit_min` is set to 7-2=5.
```

jorangreef replied:
Workings like these are always so cool.

jorangreef replied:
Remzi Arpaci-Dusseau would be proud. Just like the small workings in their OSTEP book.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -3060,10 +3138,10 @@ pub fn Replica(
                 log.debug("{}: on_request: ignoring op={} (too far ahead, checkpoint={})", .{
                     self.replica,
-                    message.header.op,
+                    self.op + 1,
```
Better this way.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -3338,6 +3416,19 @@ pub fn Replica(
+        fn op_ceiling(self: *const Self) u64 {
+            assert(self.op_checkpoint <= self.commit_min);
+            assert(self.op_checkpoint <= self.op);
+            assert((self.op_checkpoint + 1) % config.lsm_batch_multiple == 0);
```
Perhaps we could have a `checkpoint_op(op) bool` helper to encapsulate all these?

jorangreef replied:
Let's rename to `op_checkpoint_next()` since that's what it is: the next checkpoint op after `op_checkpoint`.

sentientwaffle replied:
Are you proposing that `op_checkpoint_next` replace `op_ceiling`? I think they still need to be separate because when `op_checkpoint=0`, `op_checkpoint_next == journal_slot_count - lsm_batch_multiple`, but `op_ceiling == journal_slot_count - 1` (you're right that `op_ceiling` doesn't need a branch, and has an off-by-one).

jorangreef replied:
Thanks for the explanation, I understand better now.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -3338,6 +3416,19 @@ pub fn Replica(
+            if (self.op_checkpoint == 0) {
+                return config.journal_slot_count - 1;
+            } else {
+                return self.op_checkpoint + config.journal_slot_count;
```
I think there may be an off-by-one here, since `op_checkpoint` is an index and `journal_slot_count` is a count.

Adding them together produces a count, not an index. To get the index, we need to subtract 1.

I think this also means we can get rid of the branch here, since it's then the same calculation for both branches.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -3354,13 +3445,13 @@ pub fn Replica(
         /// Returns whether the replica is the leader for the current view.
         /// This may be used only when the replica status is normal.
-        fn leader(self: *Self) bool {
+        fn leader(self: *const Self) bool {
```
Good const catch.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -439,14 +439,18 @@ pub fn Replica(
-                    assert(self.op == 0);
+                    // TODO Assert that this path isn't taken more than once.
                     self.op = self.journal.op_maximum();
+                    assert(self.op >= self.op_checkpoint);
```
Do we want to relate this also to `op_checkpoint_next()`?

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -2388,10 +2399,75 @@ pub fn Replica(
+                self.state_machine.forest.checkpoint(
+                    commit_op_checkpoint_forest_callback,
+                    commit,
+                );
```
Let's rename `commit` to `op` since that's what we call it within the forest checkpoint signature (same for compact signature etc.).

jorangreef replied:
I think we may have a bug here.

The op at which we call checkpoint is always shifting one measure early each log cycle. So we're not always checkpointing at the last slot in the journal. It's always moving left, always relative to `op_checkpoint`.

Can we simply compare op against `op_checkpoint_next()` here, to see if we should checkpoint?

jorangreef reviewed (changes requested):
Awesome to be so close to wrapping the log.

sentientwaffle committed:
VSR: `op_checkpoint_next`, `op_checkpoint_trigger`

- I misunderstood how the `op_checkpoint` moves. I had thought it would jump by `journal_slot_count` every checkpoint, but it actually falls behind a measure every time.
- Also fixes several off-by-one errors.
- Diagram the various op numbers to avoid them in the future.
- Cleanup and suggestions from code review.

kprotty commented on src/vsr/replica.zig:
```diff
@@ -3344,6 +3427,47 @@ pub fn Replica(
+
+            const op = if (self.op_checkpoint == 0)
+                // First wrap: op_checkpoint_next = 8-2-1 = 5
+                config.journal_slot_count - config.lsm_batch_multiple - 1
```
Assuming `lsm_batch_multiple = 4` here, wouldn't this be `8 - 4 - 1 = 3`?

sentientwaffle replied:
I use `lsm_batch_multiple = 2` in these examples (it is noted in this function's comment). I know it isn't a realistic value, but it is useful for illustration.

jorangreef commented on src/state_machine.zig:
```diff
@@ -344,16 +344,8 @@ pub fn StateMachineType(comptime Storage: type) type {
-            // TODO self.forest.checkpoint(op, checkpoint_callback);
-            _ = op;
-            callback(self);
+        pub fn compact(self: *StateMachine, callback: fn (*Forest) void, op: u64) void {
```
Can we change this to have the callback receive `*StateMachine`? The `Replica` should only see that, and not know about the `Forest` which is internal to `StateMachine`.

jorangreef replied:
i.e. Then we call `self.forest.compact(compact_callback, op)` and then in `compact_callback` here in `StateMachine` we then use `@fieldParentPtr` on the `forest` pointer to get back to the state machine and then we call `callback(state_machine)`.

The idea is that every callback always provides a pointer to the thing that fired the callback.

jorangreef replied:
So the callback to `StateMachine` from `Replica` should be fired by `StateMachine`, passing a pointer to itself. And then turtles all the way down into `Forest`.

jorangreef replied:
This way, a `StateMachine` does not need to use a `Forest` at all. It can take care of state internally any way it wants.

sentientwaffle replied:
Sounds good, I'll change it. Just FYI I originally used `Forest` as the target to avoid the extra callback property on the `StateMachine`.

sentientwaffle replied:
I'm also adding `StateMachine.checkpoint`, since right now the replica directly calls `replica.state_machine.forest.checkpoint(...)`.

jorangreef replied:
Great!

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -736,7 +741,7 @@ pub fn Replica(
 
             // Verify that the new request will fit in the WAL.
-            if (message.header.op >= self.op_checkpoint + config.journal_slot_count) {
+            if (message.header.op > self.op_checkpoint_trigger()) {
```
Reads well.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -2394,10 +2406,79 @@ pub fn Replica(
+            assert(self.commit_prepare.?.header.op == self.commit_min);
+
+            self.op_checkpoint = self.op_checkpoint_next();
+            assert(self.op_checkpoint == self.commit_min - config.lsm_batch_multiple);
```
I think we may trip this if checkpointing ever happens at `op_checkpoint_trigger = config.lsm_batch_multiple - 1`; this is the -1 edge case? We could do `-|` instead?

sentientwaffle replied:
I don't understand: checkpointing shouldn't trigger on `op_checkpoint_trigger == self.commit_min - config.lsm_batch_multiple` because checkpointing is triggered by comparing the `op_checkpoint_trigger == self.commit_min`.

Also,
- Right now `op_checkpoint_trigger()` is implemented as `self.op_checkpoint_next() + config.lsm_batch_multiple`.
- At this line, `self.op_checkpoint` was just advanced.
- `self.commit_min` is the "previous" `op_checkpoint_trigger()`. i.e. just before updating the `self.op_checkpoint` on line `2468`, `self.commit_min == self.op_checkpoint_trigger()`.

jorangreef replied:
At present, we would be fine.

I am thinking about if we optimize to checkpoint at the earliest possible checkpoint, for example, to accelerate startup after spin down, then for `batch_multiple=2`, we could have `op_checkpoint_trigger=1`, `commit_min=1`, `op_checkpoint_next=0` (which then stays the same i.e. 0 after the checkpoint at A):

```zig
   A  B
|01|23|45
```

And then if I understand correctly, `commit_min=1 - batch_multiple=2` would underflow the subtraction.

jorangreef replied:
We could also assert that a duplicate checkpoint like this never happens, for example, if `op_checkpoint=0` already then we shouldn't checkpoint at A because it wouldn't make a difference, since it's not helping with snapshotting the log—we're starting up reading from the same place in the log.

sentientwaffle replied:
> And then if I understand correctly, `commit_min=1 - batch_multiple=2` would underflow the subtraction.

I think the solution is to not trigger a checkpoint if nothing would actually be checkpointed.

> We could also assert that a duplicate checkpoint like this never happens, for example, if `op_checkpoint=0` already then we shouldn't checkpoint at A because it wouldn't make a difference, since it's not helping with snapshotting the log—we're starting up reading from the same place in the log.

So essentially `assert(op_checkpoint < op_checkpoint_next())`?

jorangreef replied:
Perfect.

jorangreef reviewed (commented):
Nice work on this. I took another look and think I may have found an assertion that could trip.
